### DISCUSSION TITLE
build_nested_query and then parse_nested_query gives different output

### DISCUSSION THREAD

**Discussion by doits** [rack/rack#1974] [General]
I just realized the following:

```rb
input = {"a"=>[{"name"=>"x"}, {"other"=>"y"}]}

query = Rack::Utils.build_nested_query(input)
# => "a[][name]=x&a[][other]=y"

output = Rack::Utils.parse_nested_query(query)
# => {"a"=>[{"name"=>"x", "other"=>"y"}]}

input == output
# => false
```

So it looks like not every object/hash can be encoded as a query parameter.

I'm pretty sure this is a known fact and I'm not the first one realizing this. Is there some specification or rfc that was incomplete in this part?

**Comment by jeremyevans** [comment]
This is expected.  There's no way to differentiate between `{"a"=>[{"name"=>"x"}, {"other"=>"y"}]}` and `{"a"=>[{"name"=>"x", "other"=>"y"}]}`, both result in `"a[][name]=x&a[][other]=y"` when building the query.  The latter is used when parsing.  In my apps, I generally use the hash format and then convert the hash to an array:

```ruby

input = {"a"=>[{"name"=>"x"}, {"other"=>"y"}]}

input.each do |k, v|
  input[k] = v.map.with_index{|x, i| [i.to_s, x]}.to_h
end
input
# => {"a"=>{'0'=>{"name"=>"x"}, '1'=>{"other"=>"y"}}}

query = Rack::Utils.build_nested_query(input)
# => "a[0][name]=x&a[1][other]=y"

output = Rack::Utils.parse_nested_query(query)
# => {"a"=>{"0"=>{"name"=>"x"}, "1"=>{"other"=>"y"}}}

output.each do |k, v|
   output[k] = v.sort_by{|k,| k.to_i}.map(&:last)
end
output
# => {"a"=>[{"name"=>"x"}, {"other"=>"y"}]}

input = {"a"=>[{"name"=>"x", "other"=>"y"}]}

input.each do |k, v|
  input[k] = v.map.with_index{|x, i| [i.to_s, x]}.to_h
end
input
# => {"a"=>{"0"=>{"name"=>"x", "other"=>"y"}}}

query = Rack::Utils.build_nested_query(input)
# => "a[0][name]=x&a[0][other]=y"

output = Rack::Utils.parse_nested_query(query)
# => {"a"=>{"0"=>{"name"=>"x", "other"=>"y"}}}

output.each do |k, v|
   output[k] = v.sort_by{|k,| k.to_i}.map(&:last)
end
output
# => {"a"=>[{"name"=>"x", "other"=>"y"}]}
```

**Reply by doits** [reply]
Allright, thanks for the fast response and info how to overcome it.

I have a Rails app where urls are generated which internally uses `build_nested_query` and `parse_nested_query` afaik. To follow your advice it would mean to transform query params before they are passed to Rails' helpers (like `url_for`) and transform them back after they are parsed by Rails (e.g. wrap `params` in another method).

Or is there maybe a setting for Rack (or Rails) to use this different way to build queries like the way you mentioned automatically? If I see it correctly it boils down to always using explicit indexes when building a query and transforming them back when parsing a query string, so it is "just" a different way to build and parse queries?
